design/open/hard-delete.md

```diff
@@ -0,0 +1,206 @@
+the reference counter
+
+
+#### Google Store
+
+Store the reference counter in the blob's metadata, use the `patch object` API with the `ifMetagenerationMatch` conditional.
+
+
+## 3. Tracking references in staging manager
+
+Staging manager will track physical addresses of staged objects and the staging tokens referencing them.
+We will introduce a new object to the database:
+
+      key    = <repo>/references/<physical_address>/<staging_token>  
+      value  = state enum (staged|deleted|comitted)
+               last_modified timestamp
+
+On Upload object to a branch, we will add a key in the references path with the physical address and current staging token and 
+mark it as staged.   
+On Delete object from a staging area, we will update the reference key with value `deleted`  
+On Commit, we will update the reference key with value `comitted`
+
+A background job will be responsible for scanning the references prefix, and handling the references according to the state.
+The `last_modified` parameter is used to prevent any race conditions between reference candidates for deletion and ongoing operations 
+which might add references for this physical address. The assumption is that when all references of a physical address 
+are in `deleted` state, after a certain amount of time (TBD), this physical address cannot be referenced anymore (all staging tokens were either dropped or committed).
+
+### Background delete job (pseudo-code)
+
+1. Scan `reference` prefix (can use after prefix for load balancing)
+2. For each `physical_address` read all entries
+      1. If found state == 'committed' in any entry
+         1. Delete all keys for `physical_address`, by order of state: deleted -> staged -> committed
+      2. If state == 'deleted' in all entries and min(`last_modified`) < `TBD`
+         1. Hard-delete object
+         2. Delete all keys for `physical_address`
+
+### Key Update Order of Precedence
+
+1. `commited` state takes precedence over all (terminal state - overrides any other value) - uses **Set**
+2. `deleted` state can only be done on entries which are not `committed` and uses **SetIf**
+3. `staged` state can only be done on entries which are not `committed` and uses **SetIf** 
+
+### Flows
+
+#### Upload Object
+
+1. Write blob
+2. Write reference entry to the database as staged
+3. Add staged entry to database
+
+#### Stage Object
+
+1. Object path is in the repo namespace:
+   1. Add reference entry to the database 
+2. Object path is outside the repo namespace:
+   1. lakeFS will not handle its retention
+3. Add staged entry
+
+#### LinkPhysicalAddress
+
+Assume this API uses a physical address in the repo namespace
+1. Add reference entry to the database
+2. Add staging entry
+
+#### Delete Object
+
+1. If object staged
+   1. Read reference key
+   2. If not `committed` 
+      1. Change reference state to `deleted` and update `last_modified`
+
+#### Commit
+
+1. Mark all entries in staging area as `committed`
```

Comment:
That takes some time and not easily revertable.

Reply:
I need to check further but it might be possible to perform the marking during the iteration over the changes as part of the commit process - this should save us some time

Reply:
It's a different partition. You would be replacing 2 iterations over the staging area with a single iterations that does 2 things. I don't think it would optimize the runtime.

Reply:
Or the ability to revert easily.